File: C/AnagramSearch.c
#include <stdio.h>
#include <string.h>

int main()
{
    char s1[100], s2[100], temp;

    printf("Enter two strings: ");
    scanf("%s", s1);
    scanf("%s", s2);

    int len1 = strlen(s1);
    int len2 = strlen(s2);

    // If both strings are of different length, then they are not anagrams
    if (len1 != len2)
    {
        printf("%s and %s are not anagrams\n", s1, s2);
        return 0;
    }

    // sort both string first
    for (int i = 0; i < len1 - 1; i++)
        for (int j = i + 1; j < len1; j++)
        {
            if (s1[i] > s1[j])
            {
                temp = s1[i];
                s1[i] = s1[j];
                s1[j] = temp;
            }
            if (s2[i] > s2[j])
            {
                temp = s2[i];
                s2[i] = s2[j];
                s2[j] = temp;
            }
        }

    // compare both string character by character
    for (int i = 0; i < len1; i++)
        if (s1[i] != s2[i])
        {
            puts("Strings are not anagram");
            return 0;
        }
    puts("Strings are anagrams!");

    return 0;
}
/*
Enter two strings: harsh
hrsah
Strings are anagrams!

Enter two strings: harsh
hrs
harsh and hrs are not anagrams
*/

File: Python/AnagramSearch.py
import collections


def removeWSpace(str):
    return "".join(str.split())


"""
    Checks if two strings are anagrams of each other, ignoring any whitespace.
    
    First remove any whitespace and lower all characters of both strings.
    Then create dictionaries of the counts of every character in each string.
    As well as keep a set of all characters used in both strings.
    Check to ensure every unique character are used in both strings the
        same number of times.
"""


def isAnagram(s1, s2):
    charCount1 = collections.Counter(removeWSpace(s1))
    charCount2 = collections.Counter(removeWSpace(s2))

    allChars = set(charCount1.keys())
    allChars = allChars.union(charCount2.keys())

    for c in allChars:
        if charCount1[c] != charCount2[c]:
            return False

    return True


assert isAnagram("anagram", "not a gram") == False
assert isAnagram("anagram", "na a marg") == True
assert isAnagram("William Shakespeare", "I am \t a weakish speller") == True # actually it is false
assert isAnagram("Madam Curie", "Radium came") == True
assert isAnagram("notagram", "notaflam") == False
